server/src/main/java/org/elasticsearch/rest/action/cat/RestIndicesAction.java

```diff
@@ -85,48 +86,167 @@ protected void documentation(StringBuilder sb) {
     @Override
     public RestChannelConsumer doCatRequest(final RestRequest request, final NodeClient client) {
         final String[] indices = Strings.splitStringByCommaToArray(request.param("index"));
-        final ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
-        clusterStateRequest.clear().indices(indices).metaData(true);
-        clusterStateRequest.local(request.paramAsBoolean("local", clusterStateRequest.local()));
-        clusterStateRequest.masterNodeTimeout(request.paramAsTime("master_timeout", clusterStateRequest.masterNodeTimeout()));
-        final IndicesOptions strictExpandIndicesOptions = IndicesOptions.strictExpand();
-        clusterStateRequest.indicesOptions(strictExpandIndicesOptions);
-
-        return channel -> client.admin().cluster().state(clusterStateRequest, new RestActionListener<ClusterStateResponse>(channel) {
+        final IndicesOptions indicesOptions = IndicesOptions.strictExpand();
+        final boolean local = request.paramAsBoolean("local", false);
+        final TimeValue masterNodeTimeout = request.paramAsTime("master_timeout", DEFAULT_MASTER_NODE_TIMEOUT);
+        final boolean includeUnloadedSegments = request.paramAsBoolean("include_unloaded_segments", false);
+
+        return channel -> {
+            final ActionListener<Table> listener = ActionListener.notifyOnce(new RestResponseListener<>(channel) {
+                @Override
+                public RestResponse buildResponse(final Table table) throws Exception {
+                    return RestTable.buildResponse(table, channel);
+                }
+            });
+
+            final StepListener<Map<String, Settings>> step1 = new StepListener<>();
+            sendGetIndicesRequest(indices, indicesOptions, local, masterNodeTimeout, client, step1);
+
+            final StepListener<Map<String, ClusterIndexHealth>> step2 = new StepListener<>();
+            step1.whenComplete(indicesSettings -> {
+                String[] concreteIndices = indicesSettings.keySet().toArray(Strings.EMPTY_ARRAY);
+                sendClusterHealthRequest(concreteIndices, indicesOptions, local, masterNodeTimeout, client, step2);
+            }, listener::onFailure);
+
+            final StepListener<Map<String, IndexStats>> step3 = new StepListener<>();
+            step2.whenComplete(indicesHealths -> {
+                String[] concreteIndices = indicesHealths.keySet().toArray(Strings.EMPTY_ARRAY);
+                sendIndicesStatsRequest(concreteIndices, indicesOptions, includeUnloadedSegments, client, step3);
+            }, listener::onFailure);
+
+            final StepListener<Map<String, IndexMetaData>> step4 = new StepListener<>();
```

Comment:
could step 2, 3 and 4 run in parallel?

Reply:
Yes, I used a `GroupedActionListener` for this.